security/oc_acl.c

```diff
@@ -737,55 +718,115 @@ oc_sec_ace_get_res(oc_ace_subject_type_t type, oc_ace_subject_t *subject,
-      break;
-    case OC_ACE_WC_ALL:
-      OC_DBG("Adding wildcard resource * with permission %d", permission);
-      break;
-    default:
-      break;
-    }
+  switch (res->wildcard) {
+  case OC_ACE_WC_ALL_SECURED:
+    OC_DBG("Adding wildcard resource + with permission %d", permission);
+    break;
+  case OC_ACE_WC_ALL_PUBLIC:
+    OC_DBG("Adding wildcard resource - with permission %d", permission);
+    break;
+  case OC_ACE_WC_ALL:
+    OC_DBG("Adding wildcard resource * with permission %d", permission);
+    break;
+  default:
+    break;
+  }
+#else  /* !OC_DEBUG */
+  (void)permission;
 #endif /* OC_DEBUG */
 
-    if (href) {
-      oc_new_string(&res->href, href, strlen(href));
-      OC_DBG("Adding resource %s with permission %d", href, permission);
-    }
-
-    oc_list_add(ace->resources, res);
-  } else {
-    OC_WRN("insufficient memory to add new resource to ACE");
+  if (href) {
+    oc_new_string(&res->href, href, strlen(href));
+    OC_DBG("Adding resource %s with permission %d", href, permission);
   }
-
-done:
   return res;
 }
 
+typedef struct oc_ace_res_data_t
+{
+  oc_ace_res_t *res;
+  bool created;
+} oc_ace_res_data_t;
+
+static oc_ace_res_data_t
+oc_sec_ace_get_res(oc_sec_ace_t *ace, const char *href,
+                   oc_ace_wildcard_t wildcard, uint16_t permission, bool create)
+{
+  oc_assert(ace != NULL);
+  oc_ace_res_t *res = oc_sec_ace_find_resource(NULL, ace, href, wildcard);
+  if (res) {
+    oc_ace_res_data_t data = { res, false };
+    return data;
+  }
+  if (create) {
+    res = oc_sec_add_new_ace_res(href, wildcard, permission);
+  }
+  if (!res) {
+    oc_ace_res_data_t data = { NULL, false };
+    return data;
+  }
+  oc_list_add(ace->resources, res);
+  oc_ace_res_data_t data = { res, true };
+  return data;
+}
+
+typedef struct oc_sec_ace_update_data_t
+{
+  oc_sec_ace_t *ace;
+  bool created;
+  bool created_resource;
+} oc_sec_ace_update_data_t;
+
 static bool
 oc_sec_ace_update_res(oc_ace_subject_type_t type, oc_ace_subject_t *subject,
-                      int aceid, uint16_t permission, const char *href,
-                      oc_ace_wildcard_t wildcard, size_t device)
+                      int aceid, uint16_t permission, const char *tag,
+                      const char *href, oc_ace_wildcard_t wildcard,
+                      size_t device, oc_sec_ace_update_data_t *data)
 {
-  if (oc_sec_ace_get_res(type, subject, href, wildcard, aceid, permission,
-                         device, true))
-    return true;
-  return false;
+  oc_sec_ace_t *ace =
+    oc_sec_acl_find_subject(NULL, type, subject, aceid, permission, device);
+  bool created = false;
+  if (!ace) {
+    ace = oc_sec_add_new_ace(type, subject, aceid, permission, tag, device);
+    if (!ace) {
+      return false;
+    }
+    created = true;
+  }
+  oc_ace_res_data_t res_data =
+    oc_sec_ace_get_res(ace, href, wildcard, permission, true);
+  if (res_data.res == NULL) {
+    return false;
```

Comment:
It's not leaked since it's added to the global list of ACEs by `oc_sec_add_new_ace`. I wasn't sure whether to remove the ACE or not in this case. I guess that without resources the ACE is useless, so removing probably makes more sense than keeping it.

Reply:
Fixed.